crates/grit-util/src/position.rs

```diff
@@ -1,40 +1,63 @@
-use serde::Deserialize;
-use serde::Serialize;
-use tree_sitter::Node;
-use tree_sitter::Point as TPoint;
-use tree_sitter::Range as TRange;
+use serde::{Deserialize, Serialize};
+use std::{fmt::Display, path::PathBuf};
 
-// same as TS functionality but well want to impl more functionality
-
-// note that byte if present must match the coordinates in the file.
-// in this sense maybe we want to add a file name to the struct?
-// Leave for later as currently we only operate on one file at at time.
-
-#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
 #[serde(rename_all = "camelCase")]
 
 pub struct Position {
-    // line/column are now 1-indexed
+    /// 1-based line number in the source file.
     pub line: u32,
+
+    /// 1-based column number in the source file.
     pub column: u32,
 }
 
 impl Position {
-    pub fn char_position_to_byte_position(self, context: &str) -> Self {
-        let line = self.line;
-        let column = context
-            .lines()
-            .nth(self.line as usize - 1)
-            .unwrap()
-            .chars()
-            .take(self.column as usize - 1)
-            .map(|c| c.len_utf8())
-            .sum::<usize>() as u32
-            + 1;
+    pub fn new(line: u32, column: u32) -> Self {
         Self { line, column }
     }
 
-    pub fn byte_position_to_char_position(self, context: &str) -> Self {
+    /// Returns the first position in any source string.
+    pub fn first() -> Self {
+        Self::new(1, 1)
+    }
+
+    /// Returns the last position in the given `source`.
+    pub fn last(source: &str) -> Self {
+        Self::from_byte_index(source, source.len())
+    }
+
+    /// Adds another position to this one.
+    pub fn add(&mut self, other: Position) {
+        self.line += other.line - 1;
+        self.column += other.column - 1;
+    }
+
+    /// Creates a position for the given `byte_index` in the given `source`.
+    pub fn from_byte_index(source: &str, byte_index: usize) -> Self {
+        let mut line_count = 0;
+        let mut last_line = source;
+        for line in source[..byte_index].lines() {
+            line_count += 1;
+            last_line = line;
+        }
+        let last_line_length = last_line.len() as u32;
+        Self::new(line_count, last_line_length + 1)
+    }
+
+    /// Returns the byte index for this `Position` within the given `source`.
+    pub fn byte_index(&self, source: &str) -> usize {
```

Comment:
This used to be the `get_one_indexed_position_offset()` function, but I've reimplemented it so it doesn't allocate anymore and can exit early if the position is early in the file.